File: mekaneck/virtualbrain/operators/poincare_ops.py
"""
Poincare Operators: COMPLETE, TARGET, SATISFY, EQUILIBRIUM.

Implements the Poincare computing paradigm - work backward from
completion conditions to determine trajectories.
"""

from typing import Callable, Optional, Tuple, List
from dataclasses import dataclass
import numpy as np

from ..core.types import SCoord, MentalState


@dataclass
class CompletionResult:
    """Result of a completion operation."""

    success: bool
    final_state: SCoord
    iterations: int
    trajectory: List[SCoord]
    constraint_violations: List[float]


def COMPLETE(
    partial_state: SCoord,
    constraints: List[Callable[[SCoord], float]],
    max_iterations: int = 1000,
    tolerance: float = 1e-6,
    learning_rate: float = 0.1,
) -> CompletionResult:
    """
    COMPLETE operator: Complete partial state to satisfy constraints.

    Implements Poincare completion - finding a state that satisfies
    all given constraints through gradient descent in S-entropy space.

    This is the core of backward determination: specifying the
    completion condition determines what the trajectory MUST have been.

    Args:
        partial_state: Starting state
        constraints: List of constraint functions (return 0 when satisfied)
        max_iterations: Maximum iterations
        tolerance: Convergence tolerance
        learning_rate: Initial learning rate

    Returns:
        CompletionResult with final state and trajectory
    """
    current = partial_state
    trajectory = [current]
    violations = []
    lr = learning_rate

    for iteration in range(max_iterations):
        # Compute total constraint violation
        total_violation = sum(abs(c(current)) for c in constraints)
        violations.append(total_violation)

        if total_violation < tolerance:
            return CompletionResult(
                success=True,
                final_state=current,
                iterations=iteration,
                trajectory=trajectory,
                constraint_violations=violations,
            )

        # Gradient descent step
        grad = _compute_constraint_gradient(current, constraints)

        current = current.update(
            delta_sk=-lr * grad[0],
            delta_st=-lr * grad[1],
            delta_se=-lr * grad[2],
        )
        trajectory.append(current)

        # Adaptive learning rate decay
        lr *= 0.999

    return CompletionResult(
        success=False,
        final_state=current,
        iterations=max_iterations,
        trajectory=trajectory,
        constraint_violations=violations,
    )


def TARGET(
    current: SCoord,
    target: SCoord,
    step_size: float = 0.1,
) -> SCoord:
    """
    TARGET operator: Move one step toward target state.

    Single step of gradient-based navigation toward target.

    Args:
        current: Current S-entropy coordinate
        target: Target S-entropy coordinate
        step_size: Step size for navigation

    Returns:
        Updated SCoord closer to target
    """
    grad = current.gradient(target)

    return current.update(
        delta_sk=step_size * grad[0],
        delta_st=step_size * grad[1],
        delta_se=step_size * grad[2],
    )


def SATISFY(
    state: SCoord,
    constraint: Callable[[SCoord], float],
    tolerance: float = 1e-6,
) -> bool:
    """
    SATISFY operator: Check if state satisfies constraint.

    Args:
        state: State to check
        constraint: Constraint function (returns 0 when satisfied)
        tolerance: Satisfaction tolerance

    Returns:
        True if constraint satisfied within tolerance
    """
    return abs(constraint(state)) < tolerance


def EQUILIBRIUM(
    initial_state: MentalState,
    dynamics: Callable[[MentalState, float], MentalState],
    dt: float = 0.01,
    max_time: float = 100.0,
    tolerance: float = 1e-4,
) -> Tuple[MentalState, bool, List[float]]:
    """
    EQUILIBRIUM operator: Find equilibrium state.

    Evolves system until equilibrium is reached (consciousness stabilizes)
    or max_time exceeded.

    Args:
        initial_state: Starting mental state
        dynamics: Dynamics function f(state, dt) -> new_state
        dt: Time step
        max_time: Maximum simulation time
        tolerance: Equilibrium tolerance

    Returns:
        (final_state, converged, consciousness_history)
    """
    current = initial_state
    prev_consciousness = current.consciousness
    consciousness_history = [prev_consciousness]

    t = 0.0
    while t < max_time:
        current = dynamics(current, dt)
        t += dt

        consciousness_history.append(current.consciousness)

        # Check for equilibrium
        delta_C = abs(current.consciousness - prev_consciousness)
        if delta_C < tolerance:
            return current, True, consciousness_history

        prev_consciousness = current.consciousness

    return current, False, consciousness_history


def _compute_constraint_gradient(
    state: SCoord,
    constraints: List[Callable[[SCoord], float]],
    epsilon: float = 1e-6,
) -> np.ndarray:
    """
    Compute numerical gradient of constraint violations.

    Args:
        state: Current state
        constraints: List of constraint functions
        epsilon: Finite difference step

    Returns:
        Gradient vector [d/dSk, d/dSt, d/dSe]
    """
    grad = np.zeros(3)

    for i, (delta_name, delta_val) in enumerate(
        [("delta_sk", epsilon), ("delta_st", epsilon), ("delta_se", epsilon)]
    ):
        # Perturb in positive direction
        kwargs_plus = {delta_name.replace("delta_", "delta_"): epsilon}
        if i == 0:
            state_plus = state.update(delta_sk=epsilon)
            state_minus = state.update(delta_sk=-epsilon)
        elif i == 1:
            state_plus = state.update(delta_st=epsilon)
            state_minus = state.update(delta_st=-epsilon)
        else:
            state_plus = state.update(delta_se=epsilon)
            state_minus = state.update(delta_se=-epsilon)

        # Compute gradient for each constraint
        for constraint in constraints:
            grad[i] += (constraint(state_plus) - constraint(state_minus)) / (2 * epsilon)

    return grad


def recurrence_constraint(
    initial: SCoord,
    epsilon: float = 0.1,
) -> Callable[[SCoord], float]:
    """
    Create Poincare recurrence constraint.

    The constraint is satisfied when ||state - initial|| < epsilon.

    Args:
        initial: Initial state to return to
        epsilon: Recurrence tolerance

    Returns:
        Constraint function
    """

    def constraint(state: SCoord) -> float:
        return max(0, state.distance(initial) - epsilon)

    return constraint


def equilibrium_constraint(
    target_consciousness: float,
    tolerance: float = 0.1,
) -> Callable[[MentalState], float]:
    """
    Create equilibrium constraint for consciousness level.

    Args:
        target_consciousness: Target consciousness level
        tolerance: Tolerance around target

    Returns:
        Constraint function
    """

    def constraint(state: MentalState) -> float:
        return max(0, abs(state.consciousness - target_consciousness) - tolerance)

    return constraint

File: virtual_brain/crates/vb-operators/src/poincare_ops.rs
//! Poincare Operators: COMPLETE, TARGET, SATISFY, EQUILIBRIUM.
//!
//! Implements backward constraint satisfaction for Poincare computing.

use serde::{Deserialize, Serialize};
use vb_core::types::{MentalState, SCoord};

const EPSILON: f64 = 1e-10;

/// Result of a completion operation.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CompletionResult {
    /// Whether completion succeeded
    pub success: bool,
    /// Final state after completion
    pub final_state: SCoord,
    /// Number of iterations
    pub iterations: usize,
    /// Trajectory during completion
    pub trajectory: Vec<SCoord>,
    /// Constraint violations over time
    pub constraint_violations: Vec<f64>,
}

/// COMPLETE operator: Satisfy constraints through gradient descent.
pub fn complete<F>(
    partial_state: &SCoord,
    constraints: &[F],
    max_iterations: usize,
    tolerance: f64,
    learning_rate: f64,
) -> CompletionResult
where
    F: Fn(&SCoord) -> f64,
{
    let mut state = *partial_state;
    let mut trajectory = vec![state];
    let mut violations = Vec::new();

    for iter in 0..max_iterations {
        // Compute total constraint violation
        let total_violation: f64 = constraints.iter().map(|c| c(&state).abs()).sum();
        violations.push(total_violation);

        if total_violation < tolerance {
            return CompletionResult {
                success: true,
                final_state: state,
                iterations: iter,
                trajectory,
                constraint_violations: violations,
            };
        }

        // Compute gradient of constraint violation
        let grad = compute_constraint_gradient(&state, constraints, 1e-6);

        // Update state
        state = state.update(
            -learning_rate * grad[0],
            -learning_rate * grad[1],
            -learning_rate * grad[2],
        );

        trajectory.push(state);
    }

    let final_violation: f64 = constraints.iter().map(|c| c(&state).abs()).sum();
    violations.push(final_violation);

    CompletionResult {
        success: final_violation < tolerance,
        final_state: state,
        iterations: max_iterations,
        trajectory,
        constraint_violations: violations,
    }
}

/// Compute gradient of constraint violation.
fn compute_constraint_gradient<F>(state: &SCoord, constraints: &[F], epsilon: f64) -> [f64; 3]
where
    F: Fn(&SCoord) -> f64,
{
    let violation = |s: &SCoord| -> f64 { constraints.iter().map(|c| c(s).powi(2)).sum() };

    let v0 = violation(state);

    let dsk = if state.sk + epsilon <= 1.0 {
        let s_plus = SCoord::new_unchecked(state.sk + epsilon, state.st, state.se);
        (violation(&s_plus) - v0) / epsilon
    } else {
        0.0
    };

    let dst = if state.st + epsilon <= 1.0 {
        let s_plus = SCoord::new_unchecked(state.sk, state.st + epsilon, state.se);
        (violation(&s_plus) - v0) / epsilon
    } else {
        0.0
    };

    let dse = if state.se + epsilon <= 1.0 {
        let s_plus = SCoord::new_unchecked(state.sk, state.st, state.se + epsilon);
        (violation(&s_plus) - v0) / epsilon
    } else {
        0.0
    };

    [dsk, dst, dse]
}

/// TARGET operator: Move one step toward target.
pub fn target(current: &SCoord, target: &SCoord, step_size: f64) -> SCoord {
    let dist = current.distance(target);
    if dist < EPSILON {
        return *target;
    }

    let (gsk, gst, gse) = current.gradient(target);
    let step = step_size.min(dist);

    current.update(gsk * step, gst * step, gse * step)
}

/// SATISFY operator: Check if constraint is satisfied.
pub fn satisfy<F>(state: &SCoord, constraint: F, tolerance: f64) -> bool
where
    F: Fn(&SCoord) -> f64,
{
    constraint(state).abs() < tolerance
}

/// EQUILIBRIUM operator: Find equilibrium of mental state dynamics.
pub fn equilibrium<F>(
    initial_state: &MentalState,
    dynamics: F,
    dt: f64,
    max_time: f64,
    tolerance: f64,
) -> (MentalState, bool)
where
    F: Fn(&MentalState, f64) -> MentalState,
{
    let mut state = initial_state.clone();
    let mut time = 0.0;
    let mut prev_consciousness = state.consciousness();

    while time < max_time {
        state = dynamics(&state, dt);
        time += dt;

        let new_consciousness = state.consciousness();
        if (new_consciousness - prev_consciousness).abs() < tolerance {
            return (state, true);
        }
        prev_consciousness = new_consciousness;
    }

    (state, false)
}

/// Poincare recurrence constraint.
pub fn recurrence_constraint(initial: &SCoord, epsilon: f64) -> impl Fn(&SCoord) -> f64 + '_ {
    move |s: &SCoord| s.distance(initial) - epsilon
}

/// Consciousness target constraint.
pub fn consciousness_constraint(target_c: f64, tolerance: f64) -> impl Fn(f64) -> f64 {
    move |c: f64| {
        if (c - target_c).abs() < tolerance {
            0.0
        } else {
            c - target_c
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_target() {
        let current = SCoord::origin();
        let target_s = SCoord::new(1.0, 1.0, 1.0).unwrap();

        let stepped = target(&current, &target_s, 0.1);
        assert!(stepped.distance(&current) > 0.0);
        assert!(stepped.distance(&target_s) < current.distance(&target_s));
    }

    #[test]
    fn test_satisfy() {
        let state = SCoord::new(0.5, 0.5, 0.5).unwrap();

        // Constraint: sum should be 1.5
        let constraint = |s: &SCoord| s.sk + s.st + s.se - 1.5;
        assert!(satisfy(&state, constraint, 0.01));

        // Constraint: sum should be 2.0
        let constraint2 = |s: &SCoord| s.sk + s.st + s.se - 2.0;
        assert!(!satisfy(&state, constraint2, 0.01));
    }

    #[test]
    fn test_complete() {
        let initial = SCoord::new(0.1, 0.1, 0.1).unwrap();

        // Constraint: move toward (0.5, 0.5, 0.5)
        let constraints: Vec<Box<dyn Fn(&SCoord) -> f64>> = vec![
            Box::new(|s: &SCoord| s.sk - 0.5),
            Box::new(|s: &SCoord| s.st - 0.5),
            Box::new(|s: &SCoord| s.se - 0.5),
        ];

        let result = complete(
            &initial,
            &constraints.iter().map(|c| c.as_ref()).collect::<Vec<_>>(),
            1000,
            0.01,
            0.1,
        );

        assert!(result.success);
        assert!((result.final_state.sk - 0.5).abs() < 0.05);
    }

    fn complete_with_refs<F>(
        partial_state: &SCoord,
        constraints: &[&F],
        max_iterations: usize,
        tolerance: f64,
        learning_rate: f64,
    ) -> CompletionResult
    where
        F: Fn(&SCoord) -> f64 + ?Sized,
    {
        let boxed: Vec<Box<dyn Fn(&SCoord) -> f64>> = constraints
            .iter()
            .map(|&c| {
                let c_clone = c as *const F;
                Box::new(move |s: &SCoord| unsafe { (*c_clone)(s) }) as Box<dyn Fn(&SCoord) -> f64>
            })
            .collect();

        complete(
            partial_state,
            &boxed.iter().map(|b| b.as_ref()).collect::<Vec<_>>(),
            max_iterations,
            tolerance,
            learning_rate,
        )
    }
}
